src/truncate.jl

```diff
@@ -49,33 +49,94 @@ insupport(d::Truncated{D,Union{Discrete,Continuous}}, x::Real) where {D<:Univari
 
 quantile(d::Truncated, p::Real) = quantile(d.untruncated, d.lcdf + p * d.tp)
 
-for f in [:pdf, :logpdf, :cdf, :logcdf, :ccdf, :logccdf]
-    @eval ($f)(d::Truncated, x::Int) = ($f)(d, float(x))
+function _pdf(d::Truncated, x::T) where {T<:Real}
+    if d.lower <= x <= d.upper
+        pdf(d.untruncated, x) / d.tp
+    else
+        zero(T)
+    end
+end
+
+function pdf(d::Truncated{D}, x::T) where {D<:ContinuousUnivariateDistribution, T<:Real}
+    _pdf(d, float(x))
+end
+
+function pdf(d::Truncated{D}, x::T) where {D<:DiscreteUnivariateDistribution, T<:Real}
+    isinteger(x) || zero(float(T))
+    _pdf(d, x)
+end
+
+function pdf(d::Truncated{D}, x::T) where {D<:DiscreteUnivariateDistribution, T<:Integer}
+    _pdf_(d, float(x))
+end
+
+function _logpdf(d::Truncated, x::T) where {T<:Real}
+    if (d.lower <= x <= d.upper)
+        logpdf(d.untruncated, x) - d.logtp
+    else
+        TF = float(T)
+        -TF(Inf)
+    end
+end
+
+function logpdf(d::Truncated{D}, x::T) where {D<:DiscreteUnivariateDistribution, T<:Real}
+    TF = float(T)
+    isinteger(x) || return -TF(Inf)
+    return _logpdf(d, x)
 end
 
-pdf(d::Truncated, x::T) where {T<:Real} = d.lower <= x <= d.upper ? pdf(d.untruncated, x) / d.tp : zero(T)
+function logpdf(d::Truncated{D}, x::Integer) where {D<:DiscreteUnivariateDistribution}
+    _logpdf(d, x)
+end
 
-logpdf(d::Truncated, x::T) where {T<:Real} = d.lower <= x <= d.upper ? logpdf(d.untruncated, x) - d.logtp : -T(Inf)
+function logpdf(d::Truncated{D, Continuous}, x::T) where {D<:ContinuousUnivariateDistribution, T<:Real}
```

Comment:
Why not merge this with the previous one?

Reply:
dispatch